feat(editor): Use resource locator at Simple Vector Store memory key and allow cross workflow use

Cadiac opened the pull request:
## Summary

This PR changes `Simple Vector Store` to use the Resource Locator component instead of a simple string field on the memory key picker to make selecting the same vector store on multiple nodes easier.

This PR also changes the `Simple Vector Store`'s behaviour to allowing sharing of these vector stores between workflows. Previously the memory key user picks has been prefixed on the backend with `${workflowId}__` to avoid conflicts. On the new 1.2 version of the `Simple Vector Store` this prefixing has been removed, and the given is `memoryKey` is used as is, making it possible to select the same in-memory vector store on multiple workflows.

To support creating new value on the `ResourceLocator` component support for this was added - previously something similar was done on [WorkflowSelectorParameterInput](https://github.com/n8n-io/n8n/blob/master/packages/frontend/editor-ui/src/components/WorkflowSelectorParameterInput/WorkflowSelectorParameterInput.vue) component, but now it is supported for any nodes rendering `type: resourceLocator` fields by providing the extra `typeOptions` and methods.

This clip showcases the new picker &using the in-memory vector store on another workflow.

https://github.com/user-attachments/assets/4e5a565e-a2c0-45ba-b78d-318712695a70

If user doesn't touch the key picker the default value is used - but this is now shared between workflows using the new 1.2 `Simple Vector Store` nodes. Old 1.1 versions of this node will still use workflow scoped separated vector stores with the default.

https://github.com/user-attachments/assets/06700465-e887-4887-b0f1-d22041d248f8

Using empty name for the store is treated as default value.

https://github.com/user-attachments/assets/d100ae44-8c08-4540-bf35-4fe80c85f467

Old <=1.1 `Simple Vector Store` nodes still scope the vector stores they create, but new 1.2 version vector stores can take them into use - they show up with the workflow ID prefix on the list. 1.1 version nodes cannot take vector stores created by >1.2 version nodes into use, users should upgrade their nodes instead or use the vector store created by the older node on the new one.

https://github.com/user-attachments/assets/27efe865-3d38-42c2-a05b-b827e3ec449e

## Related Linear tickets, Github issues, and Community forum posts

https://linear.app/n8n/issue/ADO-3518/memory-key-selector

## Review / Merge checklist

- [x] PR title and summary are descriptive. ([conventions](../blob/master/.github/pull_request_title_conventions.md)) <!--
   **Remember, the title automatically goes into the changelog.
   Use `(no-changelog)` otherwise.**
-->
- [ ] [Docs updated](https://github.com/n8n-io/n8n-docs) or follow-up ticket created.
- [x] Tests included. <!--
   A bug is not considered fixed, unless a test is added to prevent it from happening again.
   A feature is not complete without tests.
-->
- [ ] PR Labeled with `release/backport` (if the PR is an urgent fix that needs to be backported)

Cadiac committed:
feat(editor): Display a resource locator for picking in memory vector store memory key

Cadiac committed:
feat(editor): Make the resourceLocator support creating new resources

Cadiac committed:
refactor(editor): Bit of cleanup on the resource locator

Cadiac committed:
feat(editor): Add new simple vector store version that doesn't prefix stores with workflow id

Cadiac committed:
test(editor): Add some tests for simple vector store changes

Cadiac committed:
refactor(editor): Change how the onAddResourceClicked is defined

CharlieKolb reviewed (commented):
Nice job, code looks great!

Glad to see you considered the edge case of old versions interacting with the new one. 

Should we add a callout to the old version asking users to upgrade? We did this for ExecuteWorkflow(Trigger) since backwards compatibility was also non-trivial there: https://github.com/n8n-io/n8n/blob/8215e0b59fdd64c9efd423e8561bb2f0b61bf982/packages/nodes-base/nodes/ExecuteWorkflow/ExecuteWorkflowTrigger/ExecuteWorkflowTrigger.node.ts#L78

UX Feedback: 

I didn't really get that we're creating a new key with the name I'd be searching for, did you check with Rob on this already? Maybe we can find something better than `Search...` to encapsulate both creation and filtering.  

Some UI feedback if the key already exists would be nice at a minimum, e.g. a toast saying 'This key already exists'.
Ideally we'd include the name in the create option `Create key 'my_vector_store_key'` and/or have it disabled/greyed out if such a key already exists imo.

Cadiac commented on packages/@n8n/nodes-langchain/nodes/vector_store/shared/createVectorStoreNode/createVectorStoreNode.ts:
```diff
@@ -43,7 +43,7 @@ export const createVectorStoreNode = <T extends VectorStore = VectorStore>(
 			iconColor: args.meta.iconColor,
 			group: ['transform'],
-			version: [1, 1.1],
+			version: args.meta.version ?? [1, 1.1],
```
I do agree that this is a bit strange.

Previously all the vector stores had the same version number, which in my opinion wasn't a great solution either as they all still had node specific logic and we're bound to do fixes specific to only some of them. Here I'm touching one, but for that I didn't want to bump the versions of all the other vector store nodes - but should we just do it instead?

In my opinion all these vector store nodes should provide their own versions and there shouldn't be a common array like this here. But I didn't want to touch the other nodes right now, so I kept the existing behaviour here.

Someone adding a version `2` here shouldn't really be that difficult, they would just add it on the list of versions on the InMemory vector store too. But version 1.2 would be annoying if it had breaking changes for the in memory vector store too. Maybe the versions on all vector just have be the same

CharlieKolb replied:
I don't mind any solution here including just a comment here and in InMemory pointing this out, but very slight preference for just bumping all vector store versions instead with a comment pointing this out (in case someone looks for 1.2 specific changes in this file)

Cadiac replied:
I'll just bump them all up to 1.2.

Cadiac committed:
refactor(editor): Change all vector store nodes to have the same version

Cadiac committed:
refactor(editor): Make the add resource text better reflect what's happening, fix arrow inputs on resource locator dropdown

Cadiac commented:
Had a chat brief with Rob about UX and added the `Create key 'my_vector_store'` change, making it display the default value if empty.

I also fixed arrow key inputs on the resource locator dropdown that had been off-by-one broken for a while it seems due to the add resource option reserving index 0.

Also changed the `Search...` placeholder to `Search or create...` if the resource locator allows adding new resources.

Adding some kind of toast for "this key exists" didn't seem natural or needed now with the updating text on the create button. Hiding the button if search input matches an existing resource wasn't really feasible, as with the way resource locator is implemented there could be multiple resources with the same name, and I don't want to introduce too much logic specific to this one usage.

There's already a `notice`  callout on the component warning about its experimental prototyping nature. Since upgrading to a new version is a potentially breaking change I wouldn't push users to upgrade here, they should be fine using the old versions on their old workflows.

Cadiac committed:
fix(editor): types issue at tests, cleanup type definitions

Cadiac committed:
test(editor): fix snapshot

CharlieKolb reviewed (dismissed):
🚀

nikhilkuria commented on packages/frontend/editor-ui/src/components/ResourceLocator/ResourceLocator.vue:
```diff
@@ -346,6 +347,73 @@ const showOverrideButton = computed(
+	const addNewResourceOptions = getPropertyArgument(
+		currentMode.value,
+		'allowNewResource',
+	) as INodePropertyModeTypeOptions['allowNewResource'];
```
There is an eslint warning to remove this `as` assertion. Do you want to remove it?

nikhilkuria commented:
Great work here, @Cadiac 
Quick question, do we need to be able to create keys for all operation modes? 
<img width="485" alt="Screenshot 2025-05-20 at 11 40 15" src="https://github.com/user-attachments/assets/8753ca9e-8ef1-4d4e-8c32-cde17348284a" />

Cadiac commented:
> Great work here, @Cadiac Quick question, do we need to be able to create keys for all operation modes? <img alt="Screenshot 2025-05-20 at 11 40 15" width="485" src="https://private-user-images.githubusercontent.com/4747212/445542632-8753ca9e-8ef1-4d4e-8c32-cde17348284a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDc3MzQ4OTgsIm5iZiI6MTc0NzczNDU5OCwicGF0aCI6Ii80NzQ3MjEyLzQ0NTU0MjYzMi04NzUzY2E5ZS04ZWYxLTRkNGUtOGMzMi1jZGUxNzM0ODI4NGEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDUyMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA1MjBUMDk0OTU4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NDU5Y2ViMjExNDVjYzM0NjU0YTQ4ZTM1YWQ0YmE4YzJiYjAzMTk2MGY3NTRkY2Y4NTRiMGZiM2FiNzQ0MDJiYSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.V01dqq_-jSY3iLqQHT7XPa68vSYYl3wNPdVtnQSQ8jo">

We need to be able to pick the memory key used in all modes - whether it currently exists on the memory when user is setting this up depends on whether the node has already been run since the instance started / whether the user has created it using this picker on the other node, so I would say we should enable the same picker on all of these operation modes

Cadiac committed:
refactor(editor): remove unnecessary type cast

Cadiac committed:
Merge branch 'master' into ado-3518-memory-key-selector

Cadiac committed:
test(editor): Fix resource locator e2e tests

Cadiac committed:
test(editor): Add a note about a test that fails on dev mode

Cadiac commented:
Fixed E2E tests on this that had broken as I touched the `rlc-item` test id

shortstacked commented:
## Workflow Test Results 📊 ⚠️ **3 Warnings** (0 Failed), 👍 80 Successful out of 83 total workflows.
[View full workflow run](https://github.com/n8n-io/n8n/actions/runs/15155396938)

*Tested Ref:* `74ff3f4a840a66dd33b0f8100d76bb4b025f4c15` by @CharlieKolb
### ⚠️ Warnings (3)
| Workflow ID | Workflow Name | Reason |
|-------------|---------------|--------|
| 237 | BasicLLMChain:AzureChat | Workflow contains new data that previously did not exist. |
| 53 | ConvertKit:CustomField:create getAll update delete... | Workflow contains new data that previously did not exist. |
| 257 | Agent:auto-fix:anthropic | Workflow contains new data that previously did not exist. |

shortstacked commented:
## Workflow Test Results 📊 🔴 **1 Failed,** ⚠️ 4 Warnings, 👍 78 Successful out of 83 total workflows.
*Detail:* Workflows failing: *258*: Workflow contains 2 deleted data. [View full workflow run](https://github.com/n8n-io/n8n/actions/runs/15155609671)

*Tested Ref:* `74ff3f4a840a66dd33b0f8100d76bb4b025f4c15` by @nikhilkuria
### ❌ Failed Tests (1)
| Workflow ID | Workflow Name | Reason |
|-------------|---------------|--------|
| 258 | Agent:auto-fix:openai | Workflow contains 2 deleted data. |
### ⚠️ Warnings (4)
| Workflow ID | Workflow Name | Reason |
|-------------|---------------|--------|
| 237 | BasicLLMChain:AzureChat | Workflow contains new data that previously did not exist. |
| 35 | Slack:User:getPresence info:UserProfile:get update... | Workflow contains new data that previously did not exist. |
| 53 | ConvertKit:CustomField:create getAll update delete... | Workflow contains new data that previously did not exist. |
| 257 | Agent:auto-fix:anthropic | Workflow contains new data that previously did not exist. |

Cadiac committed:
feat(editor): Warn about in-memory vector store being shared across instance and show warning on all operation modes

shortstacked commented:
## Workflow Test Results 📊 🔴 **1 Failed,** ⚠️ 4 Warnings, 👍 78 Successful out of 83 total workflows.
*Detail:* Workflows failing: *258*: Workflow contains 2 deleted data. [View full workflow run](https://github.com/n8n-io/n8n/actions/runs/15184255930)

*Tested Ref:* `4c1d0a7aaae8c455a4dbcac68c6facb3e894d168` by @nikhilkuria
### ❌ Failed Tests (1)
| Workflow ID | Workflow Name | Reason |
|-------------|---------------|--------|
| 258 | Agent:auto-fix:openai | Workflow contains 2 deleted data. |
### ⚠️ Warnings (4)
| Workflow ID | Workflow Name | Reason |
|-------------|---------------|--------|
| 237 | BasicLLMChain:AzureChat | Workflow contains new data that previously did not exist. |
| 35 | Slack:User:getPresence info:UserProfile:get update... | Workflow contains new data that previously did not exist. |
| 53 | ConvertKit:CustomField:create getAll update delete... | Workflow contains new data that previously did not exist. |
| 257 | Agent:auto-fix:anthropic | Workflow contains new data that previously did not exist. |

janober commented:
Got released with [`n8n@1.95.0`](https://github.com/n8n-io/n8n/releases/tag/n8n%401.95.0)
